guide/src/function/error_handling.md

```diff
@@ -0,0 +1,222 @@
+
+```python
+>>> parse_int("bar")
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+ValueError: invalid digit found in string
+```
+
+As a more complete example, the following snippet defines a Rust error named `CustomIOError`. It then defines a `From<CustomIOError> for PyErr`, which returns a `PyErr` representing Python's `OSError`. Finally, it
+
+```rust
+use pyo3::exceptions::PyOSError;
+use pyo3::prelude::*;
+use std::fmt;
+
+#[derive(Debug)]
+struct CustomIOError;
+
+impl std::error::Error for CustomIOError {}
+
+impl fmt::Display for CustomIOError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Oh no!")
+    }
+}
+
+impl std::convert::From<CustomIOError> for PyErr {
+    fn from(err: CustomIOError) -> PyErr {
+        PyOSError::new_err(err.to_string())
+    }
+}
+
+pub struct Connection { /* ... */}
+
+fn bind(addr: String) -> Result<Connection, CustomIOError> {
+    if &addr == "0.0.0.0"{
+        Err(CustomIOError)
+    } else {
+        Ok(Connection{ /* ... */})
+    }
+}
+
+#[pyfunction]
+fn connect(s: String) -> Result<(), CustomIOError> {
+    bind(s)?;
+    Ok(())
+}
+
+fn main() {
+    Python::with_gil(|py| {
+        let fun = pyo3::wrap_pyfunction!(connect, py).unwrap();
+        let err = fun.call1(("0.0.0.0",)).unwrap_err();
+        assert!(err.is_instance_of::<PyOSError>(py));
+    });
+}
+```
+
+If lazy construction of the Python exception instance is desired, the
+[`PyErrArguments`]({{#PYO3_DOCS_URL}}/pyo3/trait.PyErrArguments.html)
+trait can be implemented instead of `From`. In that case, actual exception argument creation is delayed
+until the `PyErr` is needed.
+
+A final note is that any errors `E` which have a `From` conversion can be used with the `?`
+("try") operator with them. An alternative implementation of the above `parse_int` which instead returns `PyResult` is below:
+
+```rust
+use pyo3::prelude::*;
+
+fn parse_int(s: String) -> PyResult<usize> {
+    let x = s.parse()?;
+    Ok(x)
+}
+#
+# use pyo3::exceptions::PyValueError;
+#
+# fn main() {
+#     Python::with_gil(|py| {
+#         assert_eq!(parse_int(String::from("1")).unwrap(), 1);
+#         assert_eq!(parse_int(String::from("1337")).unwrap(), 1337);
+#
+#         assert!(parse_int(String::from("-1"))
+#             .unwrap_err()
+#             .is_instance_of::<PyValueError>(py));
+#         assert!(parse_int(String::from("foo"))
+#             .unwrap_err()
+#             .is_instance_of::<PyValueError>(py));
+#         assert!(parse_int(String::from("13.37"))
+#             .unwrap_err()
+#             .is_instance_of::<PyValueError>(py));
+#     })
+# }
+```
+
+## Foreign Rust error types
+
+The Rust compiler will not permit implementation of traits for types outside of the crate where the type is defined. (This is known as the "orphan rule".)
+
+Given a type `OtherError` which is defined in thirdparty code, there are two main strategies available to integrate it with PyO3:
```

Comment:
This should be `third-party` or `third party`? (same for the other usages). I'm not a native english speaker so I'm not sure 😺

Reply:
Good catch! I think `third-party` is the adjective form and `third party` the noun from looking at https://dictionary.cambridge.org/dictionary/english/third-party, so I will use `third-party` here.